Source/ACE.Database/Models/Shard/BiotaExtensions.cs

```diff
@@ -1170,9 +1171,18 @@ public static BiotaPropertiesSkill GetOrAddSkill(this Biota biota, ushort type,
                 try
                 {
                     entity = new BiotaPropertiesSkill { ObjectId = biota.Id, Type = type, Object = biota };
-                    biota.BiotaPropertiesSkill.Add(entity);
-                    skillAdded = true;
-                    return entity;
+
+                    if (SkillHelper.ValidSkills.Contains((Skill)type))
```

Comment:
This fix is not appropriate, and will break a ton of functionality.

Almost all monsters in the game will basically have 0 attack skill with this change.

Reply:
Could you elaborate at all? Do mobs still use the retired skills? 

For all skills in the valid skills hashset this should provide the same functionality as before. 

The difference is now if a call to GetOrAddSkill is made with an invalid skill (i.e. type = 0).

The first conditional...
```
var entity = biota.BiotaPropertiesSkill.FirstOrDefault(x => x.Type == type);
                if (entity != null)
                {
                    skillAdded = false;
                    return entity;
                }
```
...will be passed over because entity will be set to null (default = null?) when a skill of type 0 is not found in the biota.

The second conditional...
```
entity = new BiotaPropertiesSkill { ObjectId = biota.Id, Type = type, Object = biota };

                    if (SkillHelper.ValidSkills.Contains((Skill)type))
                    {
                        biota.BiotaPropertiesSkill.Add(entity);
                        skillAdded = true;
                        return entity;
                    }
                    else
                    {
                        skillAdded = false;
                        return entity;
                    }
```
...will populate entity with the invalid skill type = 0 and then pass over the BiotaPropertiesSkill.Add block because type = 0 is not in the valid skills hashset. Instead it will fall to the else case and set skillAdded = false and return the entity with the invalid skill type... just like it would have before these changes, except without the call to BiotaPropertiesSkill.Add and setting skillAdded to true.

thank you for the review, please help me to understand the error with this logic

Reply:
To add to the review, The BiotaExtensions should not be performing this specific type of validation. These BiotaExtensions are simply used for interfacing safely with the records themselves.

If you want to perform validation on the input, or result, it should be done by your caller or consumer of the function.

Reply:
Blindly adding skills of invalid types is considered a "safe" way to interact with the records?

You are suggesting that the validation should happen at the GetCreatureSkill level?

Reply:
"Could you elaborate at all? Do mobs still use the retired skills?"

yes, a ton of the mobs still use UnarmedCombat, for example